### ISSUE TITLE
Tooltip re-design

### ISSUE THREAD

**Issue by nickofthyme** [elastic/elastic-charts#615]
### Is your feature request related to a problem? Please describe.
As the chart functionality has grown over the past several months the current tooltip design has come to show its limitations for more complex charts.

Related to #580 

### Describe the solution you'd like
The tooltip should be redesigned to handle all of the following use cases.

#### Multiple x or other values
Up until now, the tooltip is designed with the x value being constant for all values in the tooltip. This is no longer a valid assumption. Possibilities include a table style tooltip or indented values.

**Table style**
![image](https://user-images.githubusercontent.com/19007109/78155979-c1996180-7403-11ea-8112-95eebf941abc.png)

**Indented style**
![image](https://user-images.githubusercontent.com/19007109/78157711-ea225b00-7405-11ea-868a-eb7fa58f7072.png)

The issue with the table style tooltip is that the field names for x, y, mark, etc. could be very long, creating a very wide tooltip that is less useful.

#### Aggregate values like min, max, etc.
When multiple values are allowed the logical progression would be to allow for aggregate values to be used per tooltip item. The indented files could be a better way to show this alongside the respective value, see mock up below.

![image](https://user-images.githubusercontent.com/19007109/78158135-7c2a6380-7406-11ea-88ff-c305e078dc1c.png)

#### Multiple values from the same series
![image](https://user-images.githubusercontent.com/19007109/78151908-ba238980-73fe-11ea-8f4a-93366345bf4b.png)

Up until now, the tooltip would never have two of the same series in the tooltip with the exception of the banded values.

Having multiple values in the tooltip from series based on an extension of the current tooltip design would repeat the series name for each. This seems redundant and avoidable.

#### Upper and lower (aka banded) values
Currently, we support this style but the items are repeated for each series and the order is not configurable.

![image](https://user-images.githubusercontent.com/19007109/78151560-58631f80-73fe-11ea-944b-fa54ab972698.png)

### Additional context

**Checklist**
- [x] this request is checked against already exist requests
- [x] every related Kibana issue is listed under `Kibana Cross Issues` list
- [x] `kibana cross issue` tag is associated to the issue if any kibana cross issue is present

**Comment by nickofthyme** [comment]
Also thought about a collapsed mode where if multiple values are to be displayed, they only appear in detail when hovered over a point/bar. Otherwise, they would be a simplified item, similar to what there is now.

**Comment by nickofthyme** [comment]
@cchaos would be happy to get your insight on this whenever you have the time. Thanks!

**Comment by markov00** [comment]
I'd like to link this Kibana issue here so we can think of it when redesigning the tooltip https://github.com/elastic/kibana/issues/60013#issuecomment-638194954

**Comment by miukimiu** [comment]
Yes, I can help. 🙌

**Comment by nickofthyme** [comment]
Hey @miukimiu, any update on this design?

**Comment by nickofthyme** [comment]
To add to this issue.

I think we might want to separate the cursor/crosshair styles from the tooltip styles/options.

The cursor/crosshairs should look something like...

```ts
const CrosshairType = Object.freeze({
  VerticalCursor: 'vertical' as const,
  Crosshairs: 'cross' as const,
  Follow: 'follow' as const,
  None: 'none' as const,
});
type CrosshairProps = {
  snap?: boolean;
  type?: 
}
```

And the tooltip props should be something like...

```ts
type TooltipProps = TooltipPortalSettings<'chart'> & {
  type?: 'simple' | 'complex' | 'table'; // future tooltip types
  headerFormatter?: TooltipValueFormatter;
  unit?: string;
  customTooltip?: CustomTooltip;
};
```

Doing this could allow showing the crosshairs while hiding the tooltip and vice versa.

**Comment by miukimiu** [comment]
Hi @nich07as,

I'll prioritize this and I might have designs next week.

**Comment by nickofthyme** [comment]
Tooltip should pass original `datum` within the `TooltipValue` to allow for users to extend the tooltip with additional properties outside of just the `y` value.

cc: @smith

**Comment by nickofthyme** [comment]
@miukimiu any timeline on this yet?

**Comment by sqren** [comment]
Already mentioned on Slack but posting for visibility: This is needed for https://github.com/elastic/kibana/issues/88852 which is scheduled for 7.13 🙏

**Comment by monfera** [comment]
Just linking a [slide deck](https://docs.google.com/presentation/d/1TkoPmfa4RGvWlGyyNGLZun3HpipvuAffSK9abxmvcm0/edit#slide=id.g8b713e9af1_0_19) that tries to justify (see 1st page) the increasing separation of how we gather some focus info (x,y -> glyphs -> point(s)) and what and how we render, as well as adds an example for tooltip-as-chart

**Comment by nickofthyme** [comment]
Topic: How do we handle very long tooltips. See https://github.com/elastic/kibana/pull/94247#pullrequestreview-612340782

![charts_flickering_scrollbar2](https://user-images.githubusercontent.com/1508364/111179906-cd4a4d80-85ac-11eb-8183-753d16da653c.gif)

**Comment by monfera** [comment]
The purpose of data visualization is to allow the high bandwidth sensation and ingest of information, and patterns therein, by our visual faculties. Make larger amounts of data perceivable in one glance.

The purpose of a tooltip is to provide contextual information for something that's in current focus.

Traditional tooltips consist of one or a few properties (eg. name/value) for one, or very few items.

A tooltip is a display of a subset of data, activated upon focusing on a subset of data in the original viz.

What applies to the role of original dataviz equally applies to what's shown inside the tooltip.

We wouldn't visualize an array of dozens of numbers, key/values or tuples in a table and call it a day—we would contemplate what the right form is to show the data (show top 5, and "Other"? show a viz? show a pageable/scrollable table?)

I think, equally, once the number of data points reaches the not too precise "few", a decision needs to be made:
1. should the tooltip itself show a smaller amount, eg. first 5 rows, and allow scrolling to show more if needed?
2. should it just be linked to a scrollable table that's beside the chart?
3. should it be a non-scrollable, more granular display below the chart, ie. single row detail as [here](http://martinspier.io/d3-flame-graph/) or [here](http://www.brendangregg.com/FlameGraphs/cpu-mysql-updated.svg)?
4. should the tooltip display something that is more appropriate for the contents; eg. here, it's a single layer partitioning, so it could show a treemap in a normal rectangle
5. should the tooltip only show the values for the hovered sub-bar, eg. the green rectangle?
6. if eg. only the green is in focus (FR?) should the tooltip show a single value, or a time series?
7. should the tooltip be made pinnable, to make it conducive to scrolling?
8. should the tooltip just give up and limit its height to eg. 50% of chart height? (OK it can protrude, but we don't know, by how much)

#1 is not an option, because our tooltip is activated on hover, and is dependent on what we hover over within the chart, which means, there can't be interaction over the tooltip itself, such as scroll.

I think the tooltip has the inherent problem that it occludes data; the larger it is, the more it obscures. To provide more flexibility, we can
- extract out, and separate from the tooltip, the _user gesture_ that currently leads to the tooltip activation; we can call it hover, and we should avoid calling it anything relating to tooltip; we shouldn't find "tooltip" anywhere in its source
- there are other gestures, eg. `click`, `boxSelect` etc., in theory, also linkable to a tooltip or similar
- allow our charts specification to have a `<CalloutTooltip>` spec; within this spec, the user could specify that it's activated by `hover`; and what values, with what formatter, order, etc. should be done - basically, it's like a tabular display specification (this can even be a default, eg. if the user doesn't specify or disable anything, it's still present)
- as an alternative, provide space below the chart, similar to eg. current small multiples `title` on the bottom, to render user formatted content on hover; ie. add an `<Annotation>` or some such to the spec, with an appropriate formatter, referencing `hover`
- the crosshair line becomes something analogous to `<CalloutTooltip>` - it might even show some naked value (eg. time) on the top of the chart where the crosshair line ends, where there's no graphic below it; `<CalloutTooltip>` has no idea whatsoever about `hover`, this word occurs 0 times in its code
- these things could be hooked up to `hover`, but also to eg. `click`, so it'd be possible to bring up a table or treemap with a single click on a bar, which can be hovered (I don't suggest full inception)
- the specifictions in `Chart` are responsible for enlisting the criteria (fields) that are part of the filtering aspect of the hover/tooltip; eg. should it filter the hovered time bin (bar), or the hovered country (color) or both? This gets more complex with small multiples, bubble charts and scatterplots, so I'd prefer the flexibility of just specifying _how_ the subsetting of the data is done, eg. by listing the fields that are criteria for equality filtering
- this subsetting specification technically (and internally) lives _between_ the hover gesture and the `<CalloutTooltip>` component, because one might want to drive multiple data from the same source (eg. show a crosshair in all small multiples, but show a tooltip only in the currently hovered one) - ie. the subsetting specification can be its own `Transform` spec, at least internally...
- ... and the `Transform` is responsible for things like selecting only the top geom, or all geoms that fall below the cursor, or all geoms that are fully contained inside a `boxSelect`, or all geoms that partially overlap with the `boxSelect`
- ... but to make it simple for the user, maybe they could add subsetting spec to the downstream terminal of `hover` or the upstream terminal of `<CalloutTooltip>` (or both) - such approach for shorthand notation has plenty of examples in web standards (I think, such shorthands must be superficial, ie. they should be immediately converted to the internal API that has a Transform spec, channeling data from the gesture (and screen space geom data and relevant original data) to the rendering component)

**Comment by smith** [comment]
We implemented something close to these examples for a bubble chart in APM in Kibana: https://github.com/elastic/kibana/pull/95577/files#diff-725dd22b1f6bbad4aa7a6186755643fdfd9d139de10f8ba32a2abf565af4e28a

It works, but took quite a bit of copying from the elastic-charts source and assumes those classnames continue to work. It also may have been helpful if the arguments passed through components as well as data, so I could have something like:

```ts
({ ContainerComponent, HeaderComponent, values }) => <ContainerComponent><HeaderComponent><MyCustomTooltipBody {...values} /></HeaderComponent></ContainerComponent>
```

(not that specific API, but the feature of being able to compose pieces of the tooltip.)

**Comment by miukimiu** [comment]
Thanks, @gvnmagni. This is looking great!

> I'm not fully respecting our design system, in parallel I'll be fixing this

Let me know when you reach the point of using Elastic UI, and likely, we can share some UI patterns. As @ryankeairns  pointed out in the **Product Requirement** doc there are some designs patterns for actions inside the popover that we can reuse to have some consistency (filter in, filter out..): https://github.com/elastic/eui/issues/5132.

**Comment by ryankeairns** [comment]
@gvnmagni great writeup, very thorough! The examples and prototypes paint a clear picture of where things are heading and leave me feeling excited for this to develop. Providing contextual data/actions based upon where you hover or click is resonating with me as is the scalability of the popover design you've put forth.

I look forward to participating in discussions, running through some use cases, and collaborating with the DataViz, Obs/Solution, and EUI teams.

**Comment by markov00** [comment]
Linking this other interesting possible requirement: https://github.com/elastic/kibana/issues/124034

**Comment by gvnmagni** [comment]
Thank you Chandler, I really appreciated the two examples, let me add a couple of considerations.

**Pinnable**

This option got me really interested because it would reduce the number of clicks required to get to interact with the tooltip and we will definitely have to take this in consideration. The only downside that I see is that moving through very close data points could be difficult if the tooltip stays on the interface, in your example the tooltip will cover the datapoints that live above the hovered one and to get to them we would have to move around to make the tooltip disappear and then get to the desired datapoints. This could became frustrating really fast

**Delayed-sticky**

I promise we won't do anything like this 🙂
Even though got me questioning myself why they did something like that, it really doesn't seem a good choice but there should be a reason

**Comment by cchaos** [comment]
My 2 cents on these types of interactions:

Creating a new convention around the interaction of tooltips/popovers is tricky and would need to be fully tested. I would honestly advise against it because our products need to be fast to learn and quickly aid in the user's task vs needing to learn a new interaction. I think there's a very simple solution/guideline we can adhere to that would mostly solve this issue for all usages.

### Interactive vs non-interactive tooltips:

A. **If the anchoring element is interactive itself (button, link), the tooltip cannot contain any other interactive contents.** This enforces the primary action to always be the interaction of the anchor and keeping the tooltip contents as strictly extra guidance that is not **necessary** for usability.
B. **If the anchoring element is not-interactive itself, but is strictly to open the tooltip/popover, then the tooltip can have interactive contents.** This does not mean that they **have** to click to view the tooltip, it just means that the interaction on the element forces focus onto the element which would also render the tooltip. If the user's focus then transfers to the tooltip contents, the tooltip continues to remain visible.

These guideances are very similar currently in how we differentiate between tooltips and popovers (the components themselves), with the exception of allowing popovers on hover or tooltips to contain focusable elements.

There are other ways to allow quick access to functions that may have normally been the anchor's primary action, like allowing a specific key+click combination.

The trouble with "pinnable" tooltips, is that the user still has no control over the position of this element which means it's pretty unuseful when you have multiple that overlay each other.

To circle it back to charts specifically, if we're relying on tooltips for users to see the raw values of several points at once, we're using them wrong.

**Comment by markov00** [comment]
>To circle it back to charts specifically, if we're relying on tooltips for users to see the raw values of several points at once, we're using them wrong.

Hi @cchaos can you please expand on this?

**Comment by cchaos** [comment]
Right click action should be used sparingly and only as another means to the same action. It alienates mobile users and has low discoverability. So as a shortcut to certain actions, sure, but not as the only way to that action.

**Comment by gvnmagni** [comment]
Thank you Caroline! Just wanted to confirm that it's not planned to allow users to open up multiple tooltips on the same chart, values that can be compared will take place within the same tooltip (single categories of a stacked barchart, for example).

**Comment by sqren** [comment]
> Said that, I just wanted to let you know that we are considering this and we are trying to find the best solution, thanks a lot for reaching out, if you have any suggestion please feel free to add them

@giovanniblanco  I realised I never replied. Thanks for the reply and glad to hear you are considering the options. I can see why you'd like to verify a possible solution against several use cases to make sure it's generally useable. That being said, I hope we can land on something sooner rather than later. Don't hesitate to reach out if you want to hear more about our use case.

**Comment by gvnmagni** [comment]
Hello @sqren , we actually have an interesting update! Apologies for not getting back here to notify that... we realised that the solution for this problem comes from the intrinsic modular structure of the new tooltip. In case of a timeshift, the best solution would be (from my point of view) to rely on a double header so that we can be explicit on what's represented.

In the following image you can see that approach (two ways, according to what you prefer to be your preferred dimension) applied to a case that is slightly more complex than yours since we considered 2 classic series and 2 timeshifted series, but I hope this could help you.

![time shifta](https://user-images.githubusercontent.com/18484576/215724390-7687eea6-e5de-496d-8ae8-755f35c9bc43.jpg)

This is already available and you can proceed using that, I'm tagging @markov00 that is way better than me in helping for any technical matter

**Comment by sqren** [comment]
> This is already available and you can proceed using that, 

Ahh, this is definitely a huge improvement over what we display today. I'll look into using the double header. Thank you!
